apps/dcos_l4lb/src/dcos_l4lb_mgr.erl

```diff
@@ -182,45 +182,61 @@ start_reconcile_timer() ->
 
 -spec(handle_reconcile([{key(), [backend()]}], state()) -> state()).
 handle_reconcile(VIPs, #state{route_mgr=RouteMgr, ipvs_mgr=IPVSMgr,
-        tree=Tree, nodes=Nodes, netns=Namespaces}=State) ->
+        tree=Tree, nodes=Nodes, namespaces=Namespaces}=State) ->
     % If everything is ok this function is silent and changes nothing.
     VIPs0 = vips_port_mappings(VIPs),
     VIPs1 = healthy_vips(VIPs0, Nodes, Tree),
     VIPsP = prepare_vips(VIPs1),
     Routes = get_vip_routes(VIPs1),
-    lists:foreach(fun (NetNs) ->
-        NetNsBin = netns2bin(NetNs),
-        LogPrefix = <<"netns: ", NetNsBin/binary, "; ">>,
+    lists:foreach(fun (Namespace) ->
+        NamespaceBin = namespace2bin(Namespace),
+        LogPrefix = <<"netns: ", NamespaceBin/binary, "; ">>,
 
-        PrevRoutes = get_routes(RouteMgr, NetNs),
-        DiffRoutes = dcos_net_utils:complement(Routes, PrevRoutes),
-        ok = apply_routes_diff(RouteMgr, NetNs, DiffRoutes),
-        ok = log_routes_diff(LogPrefix, DiffRoutes),
+        PrevRoutes = get_routes(RouteMgr, Namespace),
+        {RoutesToAdd, RoutesToDel} =
+            dcos_net_utils:complement(Routes, PrevRoutes),
 
-        PrevVIPsP = get_vips(IPVSMgr, NetNs),
+        PrevVIPsP = get_vips(IPVSMgr, Namespace),
         DiffVIPs = diff(PrevVIPsP, VIPsP),
-        ok = apply_vips_diff(IPVSMgr, NetNs, DiffVIPs),
-        ok = log_vips_diff(LogPrefix, DiffVIPs)
+
+        ok = remove_routes(RouteMgr, RoutesToDel, Namespace),
+        ok = log_routes_diff(LogPrefix, {[], RoutesToDel}),
+
+        ok = apply_vips_diff(IPVSMgr, Namespace, DiffVIPs),
+        ok = log_vips_diff(LogPrefix, DiffVIPs),
+
+        ok = add_routes(RouteMgr, RoutesToAdd, Namespace),
+        ok = log_routes_diff(LogPrefix, {RoutesToAdd, []})
     end, Namespaces),
     State#state{prev_ipvs=VIPs1, prev_routes=Routes}.
 
--spec(handle_vips_imp([{key(), [backend()]}], state()) -> state()).
-handle_vips_imp(VIPs, #state{route_mgr=RouteMgr, ipvs_mgr=IPVSMgr,
-        tree=Tree, nodes=Nodes, netns=Namespaces,
+-spec(handle_vips([{key(), [backend()]}], state()) -> state()).
+handle_vips(VIPs, #state{route_mgr=RouteMgr, ipvs_mgr=IPVSMgr,
```

Comment:
Is there any specific reason for a different code structure in `handle_vips` and `handle_reconcile` ?

Reply:
The logic is the same, but a bit different. In `handle_vips`, we apply all changes to all networking namespaces based on old known values. In `handle_reconcile` we check current kernel configuration and apply all changes to each networking namespaces separately.

Reply:
I understand that. My question was more on what we gain by having this difference?

Reply:
Optimization? I don't see any other reasons.